src/foam/nanos/ruler/test/RulerDAOTest.java

```diff
@@ -87,6 +92,46 @@ public void testRuleHistory(X x) {
     );
   }
 
+  public void testCompositeRuleAction(X x){
+    rule10 = (Rule) rule2.fclone();
+    rule10.setId(10);
+    //test null array of rule actions
+    CompositeRuleAction compositeAction = new CompositeRuleAction();
+    Predicate pred10 = EQ(DOT(NEW_OBJ, foam.nanos.auth.User.EMAIL), "nanos@nanos.net");
+    rule10.setPredicate(pred10);
+    rule10.setOperation(Operations.CREATE_OR_UPDATE);
+    rule10.setAction(compositeAction);
+    ruleDAO.put(rule10);
+    user1.setEmail("nanos@nanos.net");
+    user1 = (User) userDAO.put_(x, user1).fclone();
+    test(SafetyUtil.equals("nanos@nanos.net", user1.getEmail()),"no rule action resulted in no operation");
+
+    // test array of 1 action
+    RuleAction[] actions = new RuleAction[2];
+    actions[0] = (x12, obj, oldObj, ruler, agent) -> {
+      User user = (User) obj;
+      user.setEmail("action1"+user.getEmail());
+    };
+    compositeAction.setRuleActions(actions);
+    rule10.setAction(compositeAction);
+    ruleDAO.put(rule10);
+    user1.setEmail("nanos@nanos.net");
+    user1 = (User) userDAO.put_(x, user1).fclone();
+    test(SafetyUtil.equals("action1nanos@nanos.net", user1.getEmail()),"one rule action changed user email as expected: "+ user1.getEmail());
```

Comment:
user1.getEmail().equals("action1nanos@nanos.net") instead of SafetyUtil.equals?

Reply:
I think `SafetyUtil.equals` is safer because it will not throw an NPE if `user1.getEmail()` returns null.

Reply:
You can also avoid NPEs by reversing the order:
`"action1nanos@nanos.net".equals(user1.getEmail());`
I'm indifferent on the approach since it's in a test so an NPE would signal that something is broken anyway.

Do we define the `test` method somewhere? I wonder if it would be worth adding a `testEquals(Object, Object, String)` that compares the first two parameters and the third param is the error message. Not for this PR but could be useful to think about.